### PR TITLE
feat: adding the SA and NM tags

### PR DESCRIPTION
**Pull request by nh13**
Implements #45

### CODE REVIEW

**Review by jdidion — APPROVED** [review]
Few questions/suggestions - generally looks good

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 1]
```diff
@@ -718,8 +717,7 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
             let suboptimal_chain_score = chains.iter().skip(1).map(|a| a.score).max();
             match (suboptimal_chain_score, pre_alignment_score) {
                 (None, None) => None,
-                (None, Some(score)) => Some(score),
-                (Some(score), None) => Some(score),
+                (None, Some(score)) | (Some(score), None) => Some(score),
```
[the remark this answers is not in the document]
Comment (nh13) [reply fg-stitch-lib/src/align/aligners/mod.rs:720]: Me neither, AI ftw

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 2]
```diff
@@ -938,12 +944,49 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
...
-            is_first = false;
+            // Build the SA tag, then add to each record in this chain, and finally add it to the
+            // final vector of records.  Note: the SA strings must be rotated so the primary
+            // alignment is at the start
+            sa_strings.rotate_right(primary_sub_idx);
+            let sa_string = sa_strings.join(";");
+            for mut record in chain_records {
+                let data = record.data_mut();
+                data.insert(
+                    "SA".parse().unwrap(),
```
[the remark this answers is not in the document]
Comment (nh13) [reply fg-stitch-lib/src/align/aligners/mod.rs:981]: added

File: fg-stitch-lib/src/align/sub_alignment.rs [hunk 3]
```diff
@@ -15,6 +15,7 @@ pub struct SubAlignment {
     pub target_end: usize,
     pub cigar: Cigar,
     pub score: i32,
+    pub num_edits: i32,
```
[the remark this answers is not in the document]
Comment (nh13) [reply fg-stitch-lib/src/align/sub_alignment.rs:18]: changed to u32

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 4]
```diff
@@ -781,6 +779,13 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
                 subs = new_subs;
             }
 
+            // The SAM records for this chain
+            let mut chain_records = Vec::new();
+
+            // Gather the SA string for each sub-alignment so we can later rotate the list to have
+            // the primary alignment at the start
+            let mut sa_strings: Vec<String> = Vec::new();
```
[the remark this answers is not in the document]
Comment (nh13) [reply fg-stitch-lib/src/align/aligners/mod.rs:787]: If we need it elsewhere sure, but for now it's pretty self-contained.

**Pull request merged**